input/postgres/statements.go

```diff
@@ -83,25 +83,56 @@ func GetStatements(server *state.Server, logger *util.Logger, db *sql.DB, global
 	var totalTimeField string
 	var optionalFields string
 	var sourceTable string
+	var extSchema string
+	var extVersion float32 // defined in pg_stat_statements.c
+	var foundExtVersion float32
 
 	if postgresVersion.Numeric >= state.PostgresVersion13 {
-		totalTimeField = statementSQLpg13TotalTimeField
+		extVersion = 1.8
+	} else if postgresVersion.Numeric >= state.PostgresVersion95 {
+		extVersion = 1.3
+	} else if postgresVersion.Numeric >= state.PostgresVersion94 {
+		extVersion = 1.2
 	} else {
-		totalTimeField = statementSQLDefaultTotalTimeField
+		extVersion = 1.1
 	}
 
-	if postgresVersion.Numeric >= state.PostgresVersion13 {
+	err = db.QueryRow(QueryMarkerSQL+"SELECT nspname, extversion FROM pg_extension pge INNER JOIN pg_namespace pgn ON pge.extnamespace = pgn.oid WHERE pge.extname = 'pg_stat_statements'").Scan(&extSchema, &foundExtVersion)
+	if err != nil && err != sql.ErrNoRows {
+		return nil, nil, nil, err
+	}
+
+	if err == sql.ErrNoRows {
+		logger.PrintInfo("pg_stat_statements does not exist, trying to create extension...")
+		_, err = db.Exec(QueryMarkerSQL + "CREATE EXTENSION IF NOT EXISTS pg_stat_statements SCHEMA public")
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		extSchema = "public"
+		foundExtVersion = extVersion
+	}
+
+	if foundExtVersion >= 1.8 {
 		optionalFields = statementSQLpg13OptionalFields
-	} else if postgresVersion.Numeric >= state.PostgresVersion95 {
+	} else if foundExtVersion >= 1.3 {
 		optionalFields = statementSQLpg95OptionalFields
-	} else if postgresVersion.Numeric >= state.PostgresVersion94 {
+	} else if foundExtVersion >= 1.2 {
 		optionalFields = statementSQLpg94OptionalFields
 	} else {
 		optionalFields = statementSQLDefaultOptionalFields
 	}
 
-	usingStatsHelper := false
+	if foundExtVersion >= 1.8 {
+		totalTimeField = statementSQLpg13TotalTimeField
+	} else {
+		totalTimeField = statementSQLDefaultTotalTimeField
+	}
+
+	if foundExtVersion < extVersion {
```

Comment:
@seanlinsley Oh, one more thing I realize - should we only output this during test runs?

Because right now with this change we'll output it once a minute, which may be a bit frustrating.

Reply:
Is there a reason they wouldn't update the extension after first seeing the log message? I'd tend to assume most people only use the test run feature when something's broken. Hiding this message would make it likely most users wouldn't upgrade the extension, causing us to have limited access to new data.

Of course looking farther out, this type of message should be in-app instead of being hidden in the logs.

Reply:
In practice we're not really missing any data at this point (we don't rely on the newer fields), so I think it would actually be okay to not tell people all the time.

I think most people don't look at the logs (or the test output) in regular operations. But if there is a support ticket, we'd then get them to share the test output - thus the info/warning is helpful there, but I don't think it would prompt action just because its output a lot in the regular logs.

We could also push this to next week and get @uhoh-itsmaciek's opinion - not a strong preference from my end.